Allow QueueManager to operate in parallel

milljm opened the pull request:
Allow the QueueManager to utilize the threading model provided
in the base class.

Rename job_container to job throughout the Scheduler classes.

Modify the logic in all the testers to watch for a failing
test, instead of a pending test, to determine if it should set
a status.

Fix a rather nasty bug where processDownstreamTests was setting
the _failing_ test to 'skipped dependency' rather than the
downstream tests as it was supposed to.

Closes #9911

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

Commit:
Allow QueueManager to operate in parallel

Allow the QueueManager to utilize the threading model provided
in the base class.

Rename job_container to job throughout the Scheduler classes.

Modify the logic in all the testers to watch for a failing
test, instead of a pending test, to determine if it should set
a status.

Fix a rather nasty bug where processDownstreamTests was setting
the _failing_ test to 'skipped dependency' rather than the
downstream tests as it was supposed to.

Closes #9911

milljm commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -123,83 +150,84 @@ def skipPrereqs(self):
-                tester.setStatus('skipped dependency', tester.bucket_skip)
+            for failed_job in failed_jobs:
+                failed_tester = failed_job.getTester()
+                failed_tester.setStatus('skipped dependency', failed_tester.bucket_skip)
```
how was this working before!?

milljm commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -226,27 +254,27 @@ def checkRaceConditions(self, dag_object):
-                        failed_or_skipped_testers.add(tester)
+                        failed_tester = this_job.getTester()
+                        self.adjustJobStatus(this_job, failed_tester.bucket_fail, 'OUTFILE RACE CONDITION')
+                        failed_or_skipped_testers.add(failed_tester)
```
how was THIS working before!?

permcody replied:
I'm not following, what problem do you see?

milljm replied:
Earlier in this nested for loop, i was already declaring tester. And then later, I clobber it with a different tester, and started setting statuses for it.

Commit:
Fix pyflakes errors

Adjust doctring syntax in DAG contrib to match other
doctrings.

Refs #9911

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/119846/) on 3b8dfab wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/9931/site)

This comment will be updated on new commits.

permcody commented on python/TestHarness/schedulers/QueueManager.py:
```diff
@@ -504,63 +303,254 @@ def augmentQueueParamsBase(self, job_container):
+        # Set the testers output with modifications made above
+        job.setOutput(output)
+
+        # Append processResults to our stdout file
```
Either the comment, or the logic is wrong here. You say "append" but then you seek to the beginning of the fie clobbering whatever is already there.

milljm replied:
Well, I can probably word it better...
What we are trying to do here, is append the processResults output, to the file created by qsub (test_name.o12345)

```python
output = file.read(previous stdout generated by qsub)
output = tester.processResults(output)
file.seek(0)
file.write(output)
file.truncate()
file.close()
```
Keep in mind we only allow this to happen once. Now that the file is whole, with processResults output in there, we can do things like --verbose, --timing, etc...

permcody replied:
I'd just remove append then. I'd suggest saying that you are replacing the existing content with the output + status.

permcody commented on python/TestHarness/schedulers/QueueManager.py:
```diff
@@ -504,63 +303,254 @@ def augmentQueueParamsBase(self, job_container):
+
+        self.adjustJobStatus(job, bucket, json_status['caveat_message'])
+
+        third_party_job_information = {}
```
third party? You mean the PBS here right? I'm not sure this is the best name of the variable but it's probably ok

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -102,18 +102,45 @@ def killRemaining(self):
+        """ Notify derived schedulers we are finished """
+        return
+
+    def adjustJobStatus(self, job, status, caveat):
```
I'm not sure I understand why this method exists. There must be a setJobStatus() somewhere? This one does the same thing but locks? Is the caveat different too? Just trying to understand the usage of the two.

milljm replied:
I can name it setJobStatus now that you mention it. I like that name better. I named it adjust, because there is nothing stopping a tester from setting their own status themselves. But maybe that is a poor idea.

permcody replied:
👍

permcody commented on python/TestHarness/schedulers/RunParallel.py:
```diff
@@ -14,30 +14,30 @@ def validParams():
         # If we are doing recover tests
         if self.options.enable_recover and tester.specs.isValid('skip_checks') and tester.specs['skip_checks']:
-            tester.setStatus('PART1', tester.bucket_success)
+            self.adjustJobStatus(job, tester.bucket_success, 'PART1')
```
OK - so we dropped setStatus?

milljm replied:
I am using a method in the Scheduler base class to set all statuses now. adjustJobStatus sets the tester's status in a thread lock state.

permcody commented on python/contrib/dag/__init__.py:
```diff
@@ -23,8 +23,16 @@ class DAG(object):
 
+    # Added by the MOOSE group
+    def __createOriginalGraph(self):
+        """ Private method to create and protect the current state of the graph """
```
Hmm... Not sure this method always works. If you've started deleting nodes, it would give you a clone of the current state. Also, 'protect' really isn't true. Since you are returning the original graph (not a deep copy). The user could also change the cloned graph.

I think I remember you talking about the need for this method. You needed to be able to tell the graph that you were done adding nodes and edges and needed it to save it's current state. You might just call this method clone(). More work is needed if you really want this method to do the right thing when the user screws up.

milljm replied:
Yeah... I know its not really protected. This could be better designed. This was a hasty addition.

permcody commented on python/contrib/dag/__init__.py:
```diff
@@ -33,6 +41,9 @@ def add_node(self, node_name, graph=None):
         graph[node_name] = set()
 
+        # Allow original graph to grow
+        self.__original_graph = None
```
Hmm... A little bit weird.

permcody commented on python/contrib/dag/__init__.py:
```diff
@@ -69,6 +85,10 @@ def add_edge(self, ind_node, dep_node, graph=None):
             graph = self.graph
+
+        # Allow original graph to grow
+        self.__original_graph = None
```
Ah, I see what you are doing... I think a better name/method would be to call this thing `__cached_graph`. Then when you add new nodes and edges you "invalidate the cache" or set it to `None`.

permcody commented on python/contrib/dag/__init__.py:
```diff
@@ -79,18 +99,28 @@ def add_edge(self, ind_node, dep_node, graph=None):
             graph = self.graph
+
+        # save original graph
+        self.__createOriginalGraph()
```
`cacheGraph` is a good name - especially since it has nothing to do with "original"

milljm replied:
cool, will do

permcody reviewed (changes requested):
Cosmetically, these changes look OK. This class has reached a level of complexity that makes it very difficult to verify correct operation everywhere. We are really going to have to build out the test suite even more. I had a few comments but nothing huge.

milljm commented:
It will be difficult to create unit tests for this... We will have to depend on a lot of external factors. 

Although, perhaps I can create a fake session file, that contains fake tests using all the different statuses. That could work for some of QueueManagers features. I'll have to spend some time thinking on this.

permcody commented:
I really think we can create test cases for most functionality. Yes, some of it has to be creative, but you are definitely on the right track by checking in a test session file. We (you) wrote this logic to handle these cases so they must be something that we think will happen. We have a lot of control of what/how this thing works. If you need help coming up with a way to test something, just talk to me or @brianmoose.

Commit:
Change naming conventions and address comments

Refs #9911

permcody commented:
Latest commit looks good. Let's get a few test cases built!

Commit:
Add unit tests for PBS

Adding unit tests for testing PBS.
Fix a bug when supplying absolute path with --pbs options.

More tests to follow. Need to add processResults (WAITING)
functionality, which will be a little different than whats
being committed here.

Refs #9931

Commit:
More Unittests for PBS

Create more unittests...

By creating additional tests, I found some things that
needed changing elsewhere:

1. The TestHarness should only print something if there
   is something to print.

2. When the Scheduler detects a failure and attempts to
   process downstream tests, it turns out the downstream
   test was handled (printed) before the failure:

   TestB ...... skipped (skipped dependency)
   TestA ...................... FAILED (why)

   This has been fixed. However! If the failure occures
   during a PBS status check, it will be reversed again
   (because we reversed the dependencies). But I feel
   that is more acceptable because so few people will be
   using PBS anyways. Better to make it look good for
   the normal run_tests. We can fix this later... its
   extremely minor.

3. Turns out we are not using specs['no_copy'] anywhere
   So remove that code from the QueueManager along with
   the 'no_copy_pattern'. These were remnants of the
   cluster launcher.

Closes #9931

milljm commented on python/TestHarness/tests/test_QueuePBS.py:
```diff
@@ -2,45 +2,117 @@
+                # never runs jobB, but _occurres_ during jobA's iteration).
+                elif tester.isSkipped():
+                    return
+
```
This is okay, because the thread timers in play, will handle timeouts/long running and kill this job if something terrible goes wrong.

milljm commented on python/TestHarness/tests/test_QueuePBS.py:
```diff
@@ -63,100 +51,364 @@ def createHarnessAndTesters(self, test_file):
+
+                # Launch the job and wait (non-blocking)
+                harness.scheduler.queueJobs(run_jobs=[job])
+                while tester.isPending():
```
This is okay, because the Scheduler is using thread timers to handle timeouts. If we happen to hang here, it will generate a TIMEOUT error.
